nixos/github-runner: init at v2.277.1

veehaitch opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

#101426

Adds [GitHub Actions Runner](https://github.com/actions/runner) to nixpkgs. Only `x86_64-linux` is supported right now. The .NET tests are run in the check phase and pass (apart from some online tests).

Required quite a few source patches to make it work with Nix:
- Tries to obtain the hash of the latest Git commit. Patched out and replaced by passing the `rev` from `fetchFromGitHub` to the property `GitInfoCommitHash` directly.
- Updates itself if a new release is detected. This cannot be disabled (see https://github.com/actions/runner/issues/485), hence, it's patched out.
- Assumes that its root is writable which is not the case for Nix as it resolves to the read-only Nix store. Introduced an environment variable `RUNNER_ROOT` which allows changing the root directory.
- Tries to install a systemd service when registering a new runner. Also patched out.
- The original release also includes Node. Replaced by a symlink to Nix's Node.

The output derivation tries to resemble the structure of the upstream releases where it makes sense. Also, the scripts `config.sh` and `run.sh` are fully usable.

Also adds a NixOS module:
- If any of the options `name`, `url`, `runnerGroup`, `extraLabels`, `tokenFile` or its content changes, a registration is triggered. 
- Relies on systemd life cycle management and isolation where possible.

Thanks @Trundle for helping out 🙏🏻 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

veehaitch committed:
github-runner: init at 2.277.1

veehaitch committed:
nixos/github-runner: initial version

SuperSandro2000 commented on pkgs/development/tools/continuous-integration/github-runner/default.nix:
```diff
@@ -0,0 +1,259 @@
+
+    # Fix FHS path
+    substituteInPlace src/Test/L0/Util/IOUtilL0.cs \
+      --replace '/bin/ln' \
+                '${coreutils}/bin/ln'
```
```suggestion
      --replace '/bin/ln' '${coreutils}/bin/ln'
```

SuperSandro2000 commented on pkgs/development/tools/continuous-integration/github-runner/default.nix:
```diff
@@ -0,0 +1,259 @@
+  postPatch = ''
+    # Relax the version requirement
+    substituteInPlace src/global.json \
+      --replace '3.1.302' \
+                '${dotnetSdk.version}'
```
```suggestion
      --replace '3.1.302' '${dotnetSdk.version}'
```

SuperSandro2000 commented on pkgs/development/tools/continuous-integration/github-runner/default.nix:
```diff
@@ -0,0 +1,259 @@
+, openssl
+, stdenv
+, zlib
+, ...
```
```suggestion
```

r-rmcgibbo commented:
Result of `nixpkgs-review pr 116775` at 73456884 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li><a href="https://gist.github.com/315c513d4e9588aff067ce9d3373d79a">github-runner</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/development/tools/continuous-integration/github-runner/default.nix:174:3:
```
    |
174 |   checkPhase = ''
    |   ^
```
</li>
  </ul>
</details>

veehaitch committed:
nixos/github-runner: add warning if tokenFile in Nix store

veehaitch committed:
github-runner: don't accept unexpected attrs

veehaitch committed:
github-runner: formatting nits

veehaitch committed:
github-runner: add pre and post hooks to checkPhase

veehaitch committed:
nixos/github-runner: update ExecStartPre= comment

veehaitch committed:
nixos/github-runner: adapt tokenFile option description

Also note that not only a change to the option value will trigger a
reconfiguration but also modifications to the file's content.

veehaitch committed:
nixos/github-runner: remove mkDefault for DynamicUser=

veehaitch committed:
nixos/github-runner: create a parent for systemd dirs

Adds a parent directory "github-runner/" to all of the systemd lifecycle
directories StateDirectory=, RuntimeDirectory= and LogDirectory=.

Doing this has two motivations:

1. Something like this would required if we want to support multiple
   runners configurations. Please note that this is already possible
   using NixOS containers.
2. Having an additional parent directory makes it easier to remap
   any of the directories. Without a parent, systemd is going to
   complain if, for example, the given StateDirectory= is a symlink.

veehaitch committed:
nixos/github-runner: use specifier to get abs runtime path

kfollesdal commented on nixos/modules/services/continuous-integration/github-runner.nix:
```diff
@@ -0,0 +1,289 @@
+        Changing this option triggers a new runner registration.
+      '';
+      example = "nixos";
+      default = null;
```
I have testet your PR, work fine so fare :-) But get following error if I do not set option `name`. 
```
error: cannot coerce null to a string

       at /nix/store/rfxj941pv6yf3p2rikhx8wsw3y1q7m34-source/nixos/modules/services/continuous-integration/github-runner.nix:6:18:

            5|   svcName = "github-runner";
            6|   systemdUser = "${svcName}-${cfg.name}";
             |                  ^
            7|   systemdDir = "${svcName}/${cfg.name}";
```
Looks like hostname is not set as default as described. But maybe I have done something wrong. But after setting name it works fine for me.

veehaitch replied:
Good catch, thanks! Hopefully this is fixed in 9dcd2ec516fcf30f1c34afe4122a82306da5ef62.

veehaitch committed:
nixos/github-runner: use hostname as default for option `name`

Until now, the runner registration did not set the `--name` argument if
the configuration option was `null`, the default for the option.
According to GitHub's documentation, this instructs the registration
script to use the machine's hostname.

This commit causes the registration to always pass the `--name` argument
to the runner configuration script. The option now defaults to
`networking.hostName` which should be always set on NixOS.

This change becomes necessary as the systemd service name includes the
name of the runner since fcfa809 and, hence, expects it to be set. Thus,
an unset `name` option leads to an error.

veehaitch committed:
nixos/github-runner: use types.str for `name` option

Forcing a `name` option to comply with a pattern which could also be
used as a hostname is probably not required by GitHub.

veehaitch committed:
nixos/github-runner: pass dir paths explicitly for ExecStartPre=

veehaitch committed:
nixos/github-runner: update variable and script naming

veehaitch committed:
nixos/github-runner: let systemd choose the user/group

User and group naming restrictions are a complex topic [1] that I don't
even want to touch. Let systemd figure out the username and group and
reference it in our scripts through the USER environment variable.

[1] https://systemd.io/USER_NAMES/

veehaitch committed:
Revert "nixos/github-runner: use types.str for `name` option"

The escaping applied to the subdirectory paths given to StateDirectory=,
RuntimeDirectory= and LogsDirectory= apparently doesn't use the same
strategy that is used to escape unit names (cf. systemd-escape(1)). This
makes it unreasonably hard to construct reliable paths which work for
StateDirectory=/RuntimeDirectory=/LogsDirectory= and ExecStartPre=.

Against this background, I decided to (re-)apply restrictions to the
name a user might give for the GitHub runner. The pattern for
`networking.hostName` seems like a reasonable choice, also as its value
is the default if the `name` option isn't set.

This reverts commit 193ac67ba337990c22126da24a775c497dbc7e7d.

veehaitch committed:
nixos/github-runner: use types.path for `tokenFile` option

veehaitch committed:
nixos/github-runner: escape options used as shell arguments

veehaitch committed:
nixos/github-runner: wait for network-online.target

veehaitch committed:
github-runner: ignore additional online tests

domenkozar commented:
@veehaitch is this ready?

veehaitch commented:
> @veehaitch is this ready?

I'd say: yes! I've been using the runner without any issues since opening this PR. 

Obviously, there's room for improvement. Mainly support for multiple runner definitions (currently, we can achieve this with containers) and runner registration using a personal access token. I'd suggest, however, to implement those improvements in follow-ups. This PR is pretty big already.

zimbatm commented:
sweet!

mweinelt commented:
@zimbatm This causes an eval error, because the kerberos attribute isn't pased.

> anonymous function at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/development/tools/continuous-integration/github-runner/default.nix:1:1 called without required argument 'kerberos', at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/lib/customisation.nix:69:16

https://gist.github.com/GrahamcOfBorg/2f92163fa71808194f67a2b60974f5a3

domenkozar commented:
How come CI didn't catch that?

veehaitch commented:
https://github.com/NixOS/nixpkgs/commit/dcb501f99325 wasn't included yet in this branch's base

SuperSandro2000 commented:
> How come CI didn't catch that?

The ofborg results are a week old I think.

dhess commented:
Thanks, this is fantastic!
